### ISSUE TITLE
TCP: calculate bytes sent (n_send) and received (n_recv)

### ISSUE THREAD

**Issue by wilsonNg007** [microsoft/ntttcp-for-linux#77]
It looks like in the n_send and n_recv functions in tcpstream.c, there can be bytes missing from the calculation.    For example, in below n_recv function in the 1st else statement which it returns just the error code, but the error could happen after some packets have been received during the while loop, thus the number of those received packets were not returned.   Similar symptom on the n_send function as well. 

int n_recv(int fd, char *buffer, size_t total)
{
	register ssize_t rtn;
	register size_t left = total;

	while (left > 0) {
		rtn = recv(fd, buffer, left, 0);
		if (rtn < 0) {
			if (errno == EINTR || errno == EAGAIN) {
				break;
			} else {
				printf("socket read error: %d\n", errno);
				return ERROR_NETWORK_READ;
			}
		} else if (rtn == 0)
			break;

		left -= rtn;
		buffer += rtn;
	}

	return total - left;
}

**Comment by wilsonNg007** [comment]
I also noticed for tcp and udp, the__sync_fetch_and_add is used to accumulate bytes returned from the recv path, but not for the send path. I think that might be another issue.

**Comment by wilsonNg007** [comment]
@lubaihua33, @LiliDeng, @simonxiaoss  -- could you (or someone) pls help take a look?   We are using ntttcp to cover reliability stress - and one of the criteria is to ensure no bytes lost/dropped (i.e. the number of bytes sent by the sender are successfully delivered to the receiver).  Thanks.

**Comment by wilsonNg007** [comment]
@lubaihua33  Thanks!     Some follow-up questions:

1) With your experiment, did you try multiple senders/receivers and running longer time, to check if the total aggregated bytes sent and received are equal?

2) I was thinking along the same line where the receiver may be in the middle of receiving bytes and got run-time expiration and then exit without finishing the rest of recv.   So I did another experiment to specify receiver's runtime to be longer (-t 65) than the sender (-t 60).  However, the receiver ends up receiving more bytes in total than the sender has sent.  The sender has already stopped after 60 secs (from looking at the output log), so shouldn't send anything more.  The receiver continue for another 5 secs, so should continue recv until it return no more bytes.   But why there are more bytes coming in than the sender has sent?  Any idea?

3) I mentioned there were 2 issues in the code where they can potentially contribute to the mis-match.  Do you also think so?

**Comment by lubaihua33** [comment]
@wilsonNg007 
1. I didn't try multiple senders/receivers.
2. When I did an experiment test with your above commands, I found that the sender started sending a packet before starting to calculate, which means you can see the total bytes of the receiver are 1 byte more than the sender if you use my code and your commands. So I think what you hit may be caused by this. But I haven't tried to fix this.
3. I think so.
4. Yes, I think so.

**Comment by wilsonNg007** [comment]
Thanks @lubaihua33.

For #2, the receiver total bytes are way more than 1 byte.  Below is the run logs.   Sender specified 60s runtime and reported  40748084 total bytes sent. Receiver specified 65s runtime and reported 42820696 total bytes received  -  2,072,612 bytes more received than sender has sent.    What are those 2+M bytes coming from?


./ntttcp_may22  -s -m 1,*,10.231.111.24 -V --show-tcp-retrans --show-nic-packets eth2 -p 9091 -b 1 -n 1 -t 60                                                   
NTTTCP for Linux 1.4.0
---------------------------------------------------------
*** sender role
connections:                     1 X 1 X 1
cpu affinity:                    *
server address:                  10.231.111.24
domain:                          IPv4
protocol:                        TCP
server port starting at:         9091
sender socket buffer (bytes):    1
test warm-up (sec):              no
test duration (sec):             60
test cool-down (sec):            no
show system tcp retransmit:      yes
show packets for:                eth2
quiet mode:                      disabled
verbose mode:                    enabled
---------------------------------------------------------
15:24:19 DBG : user limits for maximum number of open files: soft: 32768; hard: 32768
15:24:19 DBG : Sync connection: local:60604 [socket:3] --> 10.231.111.24:9090
15:24:19 INFO: Test cycle time negotiated is: 65 seconds
15:24:19 INFO: 1 threads created
15:24:19 DBG : New connection: local:60634 [socket:4] --> 10.231.111.24:9091
15:24:19 INFO: 1 connections created in 2680 microseconds
15:24:19 INFO: Network activity progressing...
15:25:19 INFO: Test run completed.
15:25:19 INFO: Test cooldown is in progress...
15:25:24 INFO: Test cycle finished.
15:25:24 INFO: receiver exited from current test
15:25:24 INFO:  Thread  Time(s) Throughput
15:25:24 INFO:  ======  ======= ==========
15:25:24 INFO:  0        60.27   5.41Mbps
15:25:24 INFO: 1 connections tested
15:25:24 INFO: #####  Totals:  #####
15:25:24 INFO: test duration    :60.27 seconds
15:25:24 INFO: total bytes      :40748084
15:25:24 INFO:   throughput     :5.41Mbps
15:25:24 INFO:   retrans segs   :3
15:25:24 INFO: tcp retransmit:
15:25:24 INFO:   retrans_segments/sec   :0.05
15:25:24 INFO:   lost_retrans/sec       :0.00
15:25:24 INFO:   syn_retrans/sec        :0.00
15:25:24 INFO:   fast_retrans/sec       :0.00
15:25:24 INFO:   forward_retrans/sec    :0.00
15:25:24 INFO:   slowStart_retrans/sec  :0.00
15:25:24 INFO:   retrans_fail/sec       :0.00
15:25:24 INFO: total packets:
15:25:24 INFO:   tx_packets     :30113
15:25:24 INFO:   rx_packets     :9733
15:25:24 INFO: cpu cores        :40
15:25:24 INFO:   cpu speed      :2194.710MHz
15:25:24 INFO:   user           :0.74%
15:25:24 INFO:   system         :1.83%
15:25:24 INFO:   idle           :97.39%
15:25:24 INFO:   iowait         :0.00%
15:25:24 INFO:   softirq        :0.04%
15:25:24 INFO:   cycles/byte    :3390.85
15:25:24 INFO: cpu busy (all)   :103.75%
15:25:24 INFO: tcpi rtt         :5026 us
---------------------------------------------------------
windows_x86_64:/data/local/tmp/test $


./ntttcp_may22  -r -m 1,*,10.231.111.24 -V --show-tcp-retrans --show-nic-packets eth3 -p 9091 -b 1 -t 65                                                                                            <
NTTTCP for Linux 1.4.0
---------------------------------------------------------
*** receiver role
ports:                           1
cpu affinity:                    *
server address:                  10.231.111.24
domain:                          IPv4
protocol:                        TCP
server port starting at:         9091
receiver socket buffer (bytes):  1
test warm-up (sec):              no
test duration (sec):             65
test cool-down (sec):            no
show system tcp retransmit:      yes
show packets for:                eth3
quiet mode:                      disabled
verbose mode:                    enabled
---------------------------------------------------------
15:24:18 DBG : user limits for maximum number of open files: soft: 32768; hard: 32768
15:24:18 DBG : Interface:[lo]   Address: 127.0.0.1
15:24:18 DBG : Interface:[wlan0]        Address: 10.231.111.24
15:24:18 INFO: 2 threads created
15:24:18 DBG : ntttcp server is listening on 10.231.111.24:9091
15:24:18 DBG : ntttcp server is listening on 10.231.111.24:9090
15:24:19 DBG : Sync connection: 10.231.111.166:60604 --> local:9090 [socket 6]
15:24:19 DBG : New connection: 10.231.111.166:60634 --> local:9091 [socket 7]
15:24:19 INFO: Network activity progressing...
15:25:24 INFO: Test run completed.
15:25:24 INFO: Test cycle finished.
15:25:24 INFO:  Thread  Time(s) Throughput
15:25:24 INFO:  ======  ======= ==========
15:25:24 INFO:  0        65.00   5.27Mbps
15:25:24 INFO: #####  Totals:  #####
15:25:24 INFO: test duration    :65.00 seconds
15:25:24 INFO: total bytes      :42820696
15:25:24 INFO:   throughput     :5.27Mbps
15:25:24 INFO:   retrans segs   :0
15:25:24 INFO: tcp retransmit:
15:25:24 INFO:   retrans_segments/sec   :0.00
15:25:24 INFO:   lost_retrans/sec       :0.00
15:25:24 INFO:   syn_retrans/sec        :0.00
15:25:24 INFO:   fast_retrans/sec       :0.00
15:25:24 INFO:   forward_retrans/sec    :0.00
15:25:24 INFO:   slowStart_retrans/sec  :0.00
15:25:24 INFO:   retrans_fail/sec       :0.00
15:25:24 INFO: total packets:
15:25:24 INFO:   tx_packets     :10159
15:25:24 INFO:   rx_packets     :32383
15:25:24 INFO: cpu cores        :40
15:25:24 INFO:   cpu speed      :2194.711MHz
15:25:24 INFO:   user           :0.77%
15:25:24 INFO:   system         :1.80%
15:25:24 INFO:   idle           :97.40%
15:25:24 INFO:   iowait         :0.00%
15:25:24 INFO:   softirq        :0.03%
15:25:24 INFO:   cycles/byte    :3468.94
15:25:24 INFO: cpu busy (all)   :103.67%
---------------------------------------------------------
windows_x86_64:/data/local/tmp/test $
